### PR TITLE
[DO NOT MERGE] Load texture packs from the mods directory.

### PR DESCRIPTION
**Pull request by DarioSamo**
### Please note that you will need to move your mods from mods/mm to mods.

### CODE REVIEW

**Comment by GhostlyDark** [comment]
MM Reloaded
https://github.com/GhostlyDark/MM-Reloaded/releases/tag/2024-07-28

**Comment by DonelBueno** [comment]
Is there a way to make the installation portable? If not, is it planned for version 1.2?

Thanks for the amazing work =)

**Comment by Mr-Wiseguy** [comment]
> Is there a way to make the installation portable?

Yup, if you create a file named "portable.txt" next to the executable (the file contents don't matter) then recomp will run in portable mode. In portable mode, the texture packs should be placed in `mods/mm` next to the executable. For example, if your executable is in `C:\Zelda64Recompiled\Zelda64Recompiled.exe`, then you should have `C:\Zelda64Recompiled\portable.txt` and the texture pack should be placed in `C:\Zelda64Recompiled\mods\mm\`.

**Comment by DonelBueno** [comment]
> > Is there a way to make the installation portable?
> 
> Yup, if you create a file named "portable.txt" next to the executable (the file contents don't matter) then recomp will run in portable mode. In portable mode, the texture packs should be placed in `mods/mm` next to the executable. For example, if your executable is in `C:\Zelda64Recompiled\Zelda64Recompiled.exe`, then you should have `C:\Zelda64Recompiled\portable.txt` and the texture pack should be placed in `C:\Zelda64Recompiled\mods\mm\`.

Thanks, it worked perfectly. It doesn't seem to save "manual resolution" setting and the custom resolution itself, though (it also didn't save before, I'm just reporting).

**Comment by Mr-Wiseguy** [comment]
> It doesn't seem to save "manual resolution" setting and the custom resolution itself

The F1 menu is just for debug purposes, nothing in it saves by design.

**Comment by ohMoka** [comment]
For whatever reason, this PR doesn't work in Game Mode on Steam Deck, with or without loading custom textures, on either the App Image or the Linux-X64 release; it just launches to a black screen. It does work and loads custom textures just fine in Desktop Mode, however. Version 1.1.1 also works in Game Mode.

**Comment by DarioSamo** [comment]
> For whatever reason, this PR doesn't work in Game Mode on Steam Deck, with or without loading custom textures, on either the App Image or the Linux-X64 release; it just launches to a black screen. It does work and loads custom textures just fine in Desktop Mode, however. Version 1.1.1 also works in Game Mode.

I've heard such a report from other people but I'm unable to reproduce it myself. Are you capable of debugging this in any way?

Have you tried a build from `dev`? I think it's unrelated to HD Textures as the pack loading is executed after clicking on "Start game". So it can't really happen from the new functionality of this PR.

**Comment by Norzman** [comment]
I'm having the same issue on steam deck with both the linux and windows builds. They work fine in desktop mode. The windows version boots if I force it to use D3D12 though that introduces a bunch of z fighting with textures in game

**Comment by Norzman** [comment]
Just tried the dev build as recommended and that also has the black screen issue on steam deck so must be something that's changed since the 1.1.1 release. 

Was able to get it to show by launching it through lutris in game mode using the steam runtime as a work around for now. Hopefully this issue can be solved in the dev branch so this workaround isn't needed

**Comment by DarioSamo** [comment]
> Just tried the dev build as recommended and that also has the black screen issue on steam deck so must be something that's changed since the 1.1.1 release.

Yes, someone has pointed out it's a particular commit regarding changes to swap chain behavior. The commit looks innocuous at first, but it's probably running into some particular behavior the deck does when booting from game mode that gets the swap chain stuck in a loop.

> Random note is that it does seem to use more resources compared to 1.1.1 with TDP capped lower when using the HD textures. Recommend capping GPU clock to atleast 1000mhz and setting frame rate to manual in the Graphics settings

I think you're just under the false impression of the pipeline compilation that's taking a very long time for Mesa. This was better hidden away during new playthroughs. When booting into existing save files right into clock town, it seems Mesa is taking an extremely long time to optimize the pipelines for some reason I've yet to find out. I noticed it took nearly five minutes when booting into clock town for the driver to finish compilation. I don't think this is any different than the previous version, it's just less noticeable as you play through the intro and it has more time to warm those up (and plus on repeated boots, the driver has already cached it).

The other reason is the pipeline warmup list is broken for this version compared to 1.1.1 because the shader keys have changed. This will be fixed later for 1.2 once the shaders no longer change.

I've already done some local testing that indicates Mesa is having a way tougher time than it should with this and even the AMD driver on Windows. Running a SPIR-V Optimizer on my end before giving it to the driver has resulted in far lower compilation times.

So in short, please avoid the recommendation and just give it some time to warm up the pipelines. There's very little difference in performance once the pipelines are optimized, but sadly this requires workarounds from our side because the driver is running into something it hasn't seen before and taking way longer than expected to finish.

**Comment by Norzman** [comment]
Ah ok that makes sense. Yeah tried it again this morning and it seems to have calmed down.

**Comment by DarioSamo** [comment]
@Norzman @ohMoka Can you confirm whether the build from this PR fixes the black screen issue on Deck? If so, I'll merge the fix. NOTE: It won't load texture packs because it's a different branch, this is just for testing that particular issue.

https://github.com/Zelda64Recomp/Zelda64Recomp/pull/451

**Comment by Norzman** [comment]
@DarioSamo Just tested it and can confirm that it fixes the black screen issue for me

**Comment by DarioSamo** [comment]
> @DarioSamo Just tested it and can confirm that it fixes the black screen issue for me

Thanks!

**Comment by DarioSamo** [comment]
Pushed the fix to this branch.

**Comment by DarioSamo** [comment]
Added Nerrel's RC version of the pack to the list.

https://github.com/Nerrel/MMHD/releases/tag/v1.2

**Comment by DarioSamo** [comment]
RT64 has been updated with [re-spirv](https://github.com/rt64/re-spirv) which should lead to much lower shader compilation times on Vulkan, which is specially useful for the Steam Deck.

As others have mentioned in this PR, there was indeed an issue that RT64 exposed where the driver could take a pretty long time to compile pipelines. re-spirv was developed to combat this problem and has achieved pretty great results.

> On a Steam Deck, a scene that took 288 seconds to build all pipelines when passing the SPIR-V binaries and spec constants directly to the driver took only 0.43 seconds when using re-spirv. This represents a decrease in pipeline creation times of over 99.8%.

This means that players on Vulkan platforms should see their framerate stabilized much quicker than before now.

**Comment by Nohmun** [comment]
I'm not sure if this is caused by an issue or my unfamiliarity with converting images to DDS, but when using a custom texture, the Hero's Shield inventory icon in a dialogue box will appear crisp when rendering at "Original" resolution and not using any downsampling, but will appear blurred when downsampling and/or using a resolution other than "Original". This is different from how the original texture behaves, remaining crisp no matter the resolution or downsampling setting. The custom texture is the same resolution as the original. Something I think may be important, when I first used texconv to convert my textures into DDS, the shield icon was blurry even without changing sampling or rendering above "Original" resolution. I had to add the "-m 1" option to change that behavior to what I described above.

Original Resolution No Downsampling Custom Texture
01_Original_NoDownsampling

Original Resolution 4x Downsampling Custom Texture
02_Original_4xDownsampling

Original Resolution No Downsampling Original Texture
03_Original_NoDownsampling

Original Resolution 4x Downsampling Original Texture
04_Original_4xDownsampling

**Comment by DarioSamo** [comment]
@Nohmun Apologies, I completely missed your comment as I don't have notifications for this enabled.

Is your custom texture trying to basically be the same resolution as the original? If so we might need a small change to check that if you haven't actually used a bigger sized texture, the original trick should be retained to keep the textures pixelated at 240p.

Could you provide the texture you were using? And yes, at this resolution level it is probably best to avoid generating mipmaps altogether below a certain resolution.

Please open this issue in the [rt64 repository](https://github.com/rt64/rt64) so we can follow it up there.

**Comment by Nohmun** [comment]
@DarioSamo Thanks for your reply, I've made an issue at the rt64 repository here: https://github.com/rt64/rt64/issues/83#issue-2564202457

**Comment by DarioSamo** [comment]
### This pull request has been replaced by https://github.com/Zelda64Recomp/Zelda64Recomp/pull/499.

### Please note that you will need to move your mods from `mods/mm` to `mods`.
